### PR TITLE
feat(lang): Update Nix grammar & improve queries

### PR DESCRIPTION
**Pull request by nrdxp**
Update grammar and queiries to the latest upstream. Also add improved indents thanks to @oxalica.

### CODE REVIEW

File: runtime/queries/nix/highlights.scm [hunk 1]
```diff
@@ -1,81 +1,88 @@
...
-(attr_identifier) @variable.other.member
-(inherit attrs: (attrs_inherited (identifier) @variable.other.member) )
+(variable_expression (identifier) @variable)
+
+(binding
+  attrpath: (attrpath (identifier)) @property)
+
+(identifier) @property
+
+(inherit_from attrs: (inherited_attrs attr: (identifier) @property) )
```
Comment (the-mikedavis) [inline runtime/queries/nix/highlights.scm:79]: This pattern is redundant because the prior stanza (L77) captures all `identifier`s as `@property` (`@variable.other.member`) - precedence is top-to-bottom in helix. We should be able to just remove it.

File: runtime/queries/nix/indents.scm [hunk 2]
```diff
@@ -1,18 +1,40 @@
+
```
Comment (the-mikedavis) [inline runtime/queries/nix/indents.scm:1]: I think these changes can mostly be reverted. Helix only supports `@indent` and `@outdent` plus some sub-scopes on `@indent` rather than `@branch` or `@ignore`

File: runtime/queries/nix/locals.scm [hunk 3]
```diff
@@ -0,0 +1,33 @@
+;; when using @local.reference, tree-sitter seems to
```
Comment (the-mikedavis) [inline runtime/queries/nix/locals.scm:1]: These can be removed. tree-sitter has difficulty detecting locals in Nix because they're lazy in Nix and always top-to-bottom in most other languages. (See the linked issue in the comment.)

I doubt tree-sitter will get support for locals tracking that would work for Nix and if it does I expect we would need to start from scratch with the queries.

**Review by the-mikedavis — CHANGES_REQUESTED** [review]
Helix uses custom scopes (see [here](https://docs.helix-editor.com/themes.html#syntax-highlighting)), so some of the scopes here need to be renamed. For example: `@property` becomes `@variable.other.member`. The changes for booleans and numbers can be reverted except that it looks like `integer` and `float` nodes have been renamed to `integer_expression` and `float_expression`. It looks like most of the changes in the grammar came from renaming nodes to have a `_expression` suffix.

File: runtime/queries/nix/highlights.scm (hunk 1 above)
Comment (nrdxp) [reply runtime/queries/nix/highlights.scm:79]: I just copied the file directly from upstream:
https://github.com/cstrahan/tree-sitter-nix/blob/master/queries/highlights.scm

Perhaps it would be best to fix there?

**Comment by nrdxp** [comment]
@the-mikedavis, I believe I addressed all your comments with a force push.

File: runtime/queries/nix/highlights.scm (hunk 1 above)
Comment (the-mikedavis) [reply runtime/queries/nix/highlights.scm]: Queries in tree-sitter grammar repositories tend to limit themselves to the scopes given by tree-sitter-cli ([these](https://github.com/tree-sitter/tree-sitter/blob/6192d2961096bda537e8ecb20de9c9dbd506cad8/cli/src/highlight.rs#L150-L171)), so there's usually some tailoring to do to use helix-specific scopes like how we separate floats and integers. The redundant stanzas can safely be removed upstream but not most of the other custom changes we carry.

File: runtime/queries/nix/highlights.scm [hunk 4]
```diff
@@ -1,81 +1,84 @@
...
-  "with"
+  "with" 
   "assert"
+  "or"
 ] @keyword
 
 ((identifier) @variable.builtin
- (#match? @variable.builtin "^(__currentSystem|__currentTime|__nixPath|__nixVersion|__storeDir|builtins)$")
+ (#match? @variable.builtin "^(__currentSystem|__currentTime|__nixPath|__nixVersion|__storeDir|builtins|false|null|true)$")
  (#is-not? local))
```
Comment (the-mikedavis) [inline runtime/queries/nix/highlights.scm:18]: could we revert the change here and on L36-L39? I think the `@constant.builtin(.boolean)` captures work better on true/false/null

File: runtime/queries/nix/highlights.scm [hunk 5]
```diff
@@ -1,81 +1,84 @@
...
   operator: _ @operator)
 
-(attr_identifier) @variable.other.member
-(inherit attrs: (attrs_inherited (identifier) @variable.other.member) )
+(variable_expression (identifier) @variable)
+
+(binding
+  attrpath: (attrpath (identifier)) @variable.other.member)
+
+(identifier) @variable.other.member
```
Comment (the-mikedavis) [inline runtime/queries/nix/highlights.scm:75]: Oh actually I didn't see the `(identifier) @variable` at the bottom of the file. I think we can just remove this one even though it would be a departure from the upstream highlights.

File: runtime/queries/nix/indents.scm [hunk 6]
```diff
@@ -1,18 +1,25 @@
...
+  (binary)
+  (has_attr)
+  (select)
+  (app)
 ] @indent
 
-[
-  "}"
-  "]"
-] @outdent
```
Comment (the-mikedavis) [inline runtime/queries/nix/indents.scm:18]: We should probably keep these unless the injects work better without them

File: runtime/queries/nix/indents.scm [hunk 7]
```diff
@@ -1,18 +1,25 @@
...
+; Conditional
+(if) @indent
```
Comment (the-mikedavis) [inline runtime/queries/nix/indents.scm:25]: This can be moved up into the `[ ] @indent` stanza above

File: runtime/queries/nix/indents.scm (hunk 6 above)
Comment (nrdxp) [reply runtime/queries/nix/indents.scm:18]: Not sure exactly why tbh, but I was getting extra indents with the previous definitions. These ones seem to work more accurately

File: runtime/queries/nix/highlights.scm (hunk 5 above)
Comment (nrdxp) [reply runtime/queries/nix/highlights.scm:75]: are you saying to remove all 4 lines here, or just line 75?

Comment (the-mikedavis) [reply runtime/queries/nix/highlights.scm:75]: Just L75. That `binding` capture above looks good

**Comment by nrdxp** [comment]
@the-mikedavis I think I was able to resolve everything. I just recently stumbled on this comment though: https://github.com/cstrahan/tree-sitter-nix/issues/13#issuecomment-1072242522 and now I'm wondering if I should rebase on oxalinca's fork to add his injections. Although what I'd really like to do is get his changes incooperated into upstream properly.

I think I'll give this a shot later. Maybe we should hold off on merging for a bit just in case I'm successful?

**Review by the-mikedavis — APPROVED** [review]

**Comment by the-mikedavis** [comment]
I think this is a good increment even without shell injections - that can always be a follow-up PR. Would you mind rebasing?

**Comment by nrdxp** [comment]
I went ahead and force pushed, and added this back in the latest force push:
```diff
diff --git a/languages.toml b/languages.toml
index 68f2f89f..fdf0513a 100644
--- a/languages.toml
+++ b/languages.toml
@@ -439,7 +439,7 @@ indent = { tab-width = 2, unit = "  " }
 
 [[grammar]]
 name = "nix"
-source = { git = "https://github.com/cstrahan/tree-sitter-nix", rev = "470b15a60520ff7b86f51732b8d8f1118c86041e" }
+source = { git = "https://github.com/cstrahan/tree-sitter-nix", rev = "6b71a810c0acd49b980c50fc79092561f7cee307" }
 
 [[language]]
 name = "ruby"
diff --git a/runtime/queries/nix/highlights.scm b/runtime/queries/nix/highlights.scm
index d15fb9c2..a171a39b 100644
--- a/runtime/queries/nix/highlights.scm
+++ b/runtime/queries/nix/highlights.scm
@@ -72,9 +72,15 @@
 (binary_expression
   operator: _ @operator)
 
+(variable_expression (identifier) @variable)
+
 (binding
   attrpath: (attrpath (identifier)) @variable.other.member)
 
+(identifier) @variable.other.member
+
+(inherit_from attrs: (inherited_attrs attr: (identifier) @variable.other.member) )
+
 [
   ";"
   "."
```

Without it, I experienced a regression in `path` type highlighting.

**Comment by the-mikedavis** [comment]
The

```
(variable_expression (identifier) @variable)
```

stanza is a good catch: it highlights variables in inherit-from statements:

```nix
{
  inherit (common) pkgs;
  #         ^ variable
  #                 ^ variable.other.member
}
```

The

```
(identifier) @variable.other.member
```

stanza takes priority over the

```
(identifier) @variable
```

on the last line, so let's remove the `(identifier) @variable` one since it's now redundant - the highlights look much better with the `@variable.other.member` capture.

The

```
(inherit_from attrs: (inherited_attrs attr: (identifier) @variable.other.member))
```

doesn't appear to have an effect because of the `(identifier) @variable.other.member` above it, I think we're safe to remove it. Based on the old highlights, it looks like the recent refactor upstream combined the old `attr_identifier` and `identifier` nodes which is what I think led to the redundant stanzas in the upstream queries like the above one. I'll make a PR to remove the redundant ones upstream.

By path highlights did you mean like `./shell.nix` in `import ./shell.nix` or in attribute paths? Those stanzas shouldn't interact with the file-system path highlights so I think you may have seen https://github.com/helix-editor/helix/issues/2484 if that's what you're talking about.

**Review by the-mikedavis — APPROVED** [review]
Looks great, thanks for working on this!

**Comment by nrdxp** [comment]
Thanks for the feedback. Also, just to confirm, it is #2484 that is the problem. After opening the same file several times I noticed sometimes paths are highlighted correcly, and sometimes they are just hightlighted as regular strings.
